데이터 분석을 위한 전처리...
데이터프레임 생성하는거 복습, loc, iloc 등 다양한 조회 실습
데이터의 타입을 바꾸고 인덱스 리셋하고 하는 것 등

# 1. 리스트로부터 데이터프레임 생성 (복습)

In [1]:
import pandas as pd

In [2]:
productList = ['ST0001','ST0002','ST0003']
yearweekList = ['202001','202002','202003']
qtyList = ['10','12','15']

In [3]:
pd.DataFrame(zip(productList,yearweekList, qtyList),
             columns=['product','yearweek','qty'],
             index=['one','two','three'])

product yearweek qty
one    ST0001   202001  10
two    ST0002   202002  12
three  ST0003   202003  15

In [6]:
# 방법 1. 여러개의 List를 zip으로 묶어서 데이터프레임 만들기.
pd.DataFrame(zip(productList,yearweekList,qtyList),
            columns=['product','yearweek','qty'])

product yearweek qty
0  ST0001   202001  10
1  ST0002   202002  12
2  ST0003   202003  15

In [19]:
# 방법 2. 딕셔너리로 묶어서 데이터프레임 만들기.
pd.DataFrame({'product':productList,'yearweek':yearweekList,'qty':qtyList})

product yearweek qty
0  ST0001   202001  10
1  ST0002   202002  12
2  ST0003   202003  15

In [17]:
# 방법 3. 각각 데이터프레임 만든 다음 concat으로 합치기.
productDF = pd.DataFrame(productList, columns=['product'])
yearweekDF = pd.DataFrame(yearweekList, columns=['yearweek'])
qtyDF = pd.DataFrame(qtyList, columns=['qty'])
pd.concat([productDF,yearweekDF,qtyDF], axis=1)

product yearweek qty
0  ST0001   202001  10
1  ST0002   202002  12
2  ST0003   202003  15

In [4]:
# 각각의 데이터를 DF로 만들고, pd.concat을 사용하는게 정석이다. 더 다양한 작업을 할 수 있다.
# 단, 인덱스가 다를 경우 초기화 해서 붙여줘야 하기 때문에 귀찮아서 크롤링에서는 zip을 많이 사용한다.

In [5]:
productDf = pd.DataFrame(productList, columns=['product'])
yearweekDf = pd.DataFrame(yearweekList, columns=['yearweek'])
qtyDf = pd.DataFrame(qtyList, columns=['qty'])
result = pd.concat([productDf,yearweekDf,qtyDf], axis=1)
result

product yearweek qty
0  ST0001   202001  10
1  ST0002   202002  12
2  ST0003   202003  15

### 인덱스가 맞지 않는다면???

In [2]:
productList = ['ST0001','ST0002','ST0003','ST0004']
yearweekList = ['202001','202002','202003']
qtyList = ['10','12','15']

In [3]:
pd.DataFrame(zip(productList,yearweekList, qtyList),
             columns=['product','yearweek','qty'],
             index=['one','two','three'])

product yearweek qty
one    ST0001   202001  10
two    ST0002   202002  12
three  ST0003   202003  15

In [4]:
# 방법 1. 여러개의 List를 zip으로 묶어서 데이터프레임 만들기.
pd.DataFrame(zip(productList,yearweekList,qtyList),
            columns=['product','yearweek','qty'])

product yearweek qty
0  ST0001   202001  10
1  ST0002   202002  12
2  ST0003   202003  15

In [5]:
# 방법 2. 딕셔너리로 묶어서 데이터프레임 만들기.
pd.DataFrame({'product':productList,'yearweek':yearweekList,'qty':qtyList})

ValueError: arrays must all be same length

In [6]:
# 방법 3. 각각 데이터프레임 만든 다음 concat으로 합치기.
productDF = pd.DataFrame(productList, columns=['product'])
yearweekDF = pd.DataFrame(yearweekList, columns=['yearweek'])
qtyDF = pd.DataFrame(qtyList, columns=['qty'])
pd.concat([productDF,yearweekDF,qtyDF], axis=1)

product yearweek  qty
0  ST0001   202001   10
1  ST0002   202002   12
2  ST0003   202003   15
3  ST0004      NaN  NaN

# 2. 데이터 조회하기 (고급)

In [2]:
import pandas as pd

url = '../exampleCode/dataset/customerdata.csv'

customerData = pd.read_csv(url)
customerData

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0    A13566  4273.900000    3            6    1.679181  Big-Screen-lover
1    A14219  3642.441950    2            4    2.682023      Sleeping-dog
2    A15312  3653.884565    2            5    3.208202      Sleeping-dog
3    A16605  3713.211107    2            6    0.900000        Early-bird
4    B10634  3391.074215    2            4    2.453656      Sleeping-dog
..      ...          ...  ...          ...         ...               ...
145  Z13253  3678.800000    2            5    2.463670  Big-Screen-lover
146  Z13534  3662.437527    3            6    0.900000        Early-bird
147  Z16428  3516.500000    2            5    2.371301  Big-Screen-lover
148  Z16735  3300.100000    2            5    1.704942  Big-Screen-lover
149  Z19835  3326.526183    2            6    1.200000        Early-bird

[150 rows x 6 columns]

### 데이터 정보 보기 .info()

In [3]:
customerData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   CUSTID       150 non-null    object 
 1   AVGPRICE     150 non-null    float64
 2   EMI          150 non-null    int64  
 3   DEVICECOUNT  150 non-null    int64  
 4   PRODUCTAGE   150 non-null    float64
 5   CUSTTYPE     150 non-null    object 
dtypes: float64(2), int64(2), object(2)
memory usage: 7.2+ KB


이를 통해 알 수 있는것은?<br>150개의 rows가 있고 0 - 149다, 컬럼의 정보는 다음과 같고 데이터 타입은 무엇이며, 150개 모두 null 값이 없다. 

In [4]:
customerData.dtypes

CUSTID          object
AVGPRICE       float64
EMI              int64
DEVICECOUNT      int64
PRODUCTAGE     float64
CUSTTYPE        object
dtype: object

In [25]:
mydict = [{'a': 1, 'b': 2, 'c': 3, 'd': 4},
          {'a': 100, 'b': 200, 'c': 300, 'd': 400},
          {'a': 1000, 'b': 2000, 'c': 3000, 'd': 4000 }]
df = pd.DataFrame(mydict)
df

a     b     c     d
0     1     2     3     4
1   100   200   300   400
2  1000  2000  3000  4000

In [75]:
mydict = [[1,2,3,4],
          [100,200,300,400],
          [1000,2000,3000,4000]]
df = pd.DataFrame(mydict)
df

0     1     2     3
0     1     2     3     4
1   100   200   300   400
2  1000  2000  3000  4000

In [40]:
mydict = [[1,2,3,4],
          [100,200,300,400],
          [1000,2000,3000,4000]]
df = pd.DataFrame(mydict, columns=['a','b','c','d'], index=['one','two','three'])
df

a     b     c     d
one       1     2     3     4
two     100   200   300   400
three  1000  2000  3000  4000

In [67]:
df

a     b     c     d
one       1     2     3     4
two     100   200   300   400
three  1000  2000  3000  4000

In [88]:
mydict = [[1,2,3,4],
          [100,200,300,400],
          [1000,2000,3000,4000]]
df = pd.DataFrame(mydict)
df

0     1     2     3
0     1     2     3     4
1   100   200   300   400
2  1000  2000  3000  4000

In [89]:
df[3]

0       4
1     400
2    4000
Name: 3, dtype: int64

In [90]:
df.loc[df[3]>2,]

0     1     2     3
0     1     2     3     4
1   100   200   300   400
2  1000  2000  3000  4000

In [91]:
df.loc[df[3]>4,]

0     1     2     3
1   100   200   300   400
2  1000  2000  3000  4000

In [92]:
mydict = [[1,2,3,4],
          [100,200,300,400],
          [1000,2000,3000,4000]]
df = pd.DataFrame(mydict, columns=['a','b','c','d'], index=['one','two','three'])
df

a     b     c     d
one       1     2     3     4
two     100   200   300   400
three  1000  2000  3000  4000

In [93]:
df['c']

one         3
two       300
three    3000
Name: c, dtype: int64

In [95]:
df.loc[df['c']>2,]

a     b     c     d
one       1     2     3     4
two     100   200   300   400
three  1000  2000  3000  4000

In [96]:
df.loc[df['c']>3,]

a     b     c     d
two     100   200   300   400
three  1000  2000  3000  4000

In [97]:
df.loc[df['c'] == 300]

a    b    c    d
two  100  200  300  400

In [101]:
df.loc[:,'d']

one         4
two       400
three    4000
Name: d, dtype: int64

In [126]:
df = pd.DataFrame([[1, 2], [4, 5], [7, 8]],
                  index=['cobra', 'viper', 'sidewinder'],
                  columns=['max_speed', 'shield'])
df

max_speed  shield
cobra               1       2
viper               4       5
sidewinder          7       8

In [104]:
df.loc[df['shield'] > 6]

max_speed  shield
sidewinder          7       8

In [105]:
df.loc[df['shield'] > 6, ['max_speed']]    # [row를 제한, column을 제한]하는데, row를 제한할 때 column명을 이용해 Series를 뽑아내고 거기에 조건을 걸어 row를 제한할 수도 있다.

max_speed
sidewinder          7

In [110]:
df.loc[['sidewinder'],['max_speed']]   # iloc에서 했던 것과 동일하게 직접 row, column을 제한할 수 있다.

max_speed
sidewinder          7

In [107]:
df.loc[lambda df: df['shield'] == 8]

max_speed  shield
sidewinder          7       8

In [119]:
# 조건에 맞는 row에 값 집어넣기
df.loc['cobra'] = 10
df

max_speed  shield
cobra              10      10
viper               4      50
sidewinder          7      50

In [120]:
# 조건에 맞는 column에 값 집어넣기
df.loc[:,['shield']] = 20
df

max_speed  shield
cobra              10      20
viper               4      20
sidewinder          7      20

In [ ]:
#########################################################################

In [121]:
# 조건에 맞는 행렬에 값 집어넣기
df.loc[['viper', 'sidewinder'], ['shield']] = 50
df

max_speed  shield
cobra              10      20
viper               4      50
sidewinder          7      50

In [ ]:

**Getting values on a DataFrame with an index that has integer labels**

Another example using integers for the index

>>> df = pd.DataFrame([[1, 2], [4, 5], [7, 8]],
...      index=[7, 8, 9], columns=['max_speed', 'shield'])
>>> df
   max_speed  shield
7          1       2
8          4       5
9          7       8

Slice with integer labels for rows. As mentioned above, note that both
the start and stop of the slice are included.

>>> df.loc[7:9]
   max_speed  shield
7          1       2
8          4       5
9          7       8

**Getting values with a MultiIndex**

A number of examples using a DataFrame with a MultiIndex

>>> tuples = [
...    ('cobra', 'mark i'), ('cobra', 'mark ii'),
...    ('sidewinder', 'mark i'), ('sidewinder', 'mark ii'),
...    ('viper', 'mark ii'), ('viper', 'mark iii')
... ]
>>> index = pd.MultiIndex.from_tuples(tuples)
>>> values = [[12, 2], [0, 4], [10, 20],
...         [1, 4], [7, 1], [16, 36]]
>>> df = pd.DataFrame(values, columns=['max_speed', 'shield'], index=index)
>>> df
                     max_speed  shield
cobra      mark i           12       2
           mark ii           0       4
sidewinder mark i           10      20
           mark ii           1       4
viper      mark ii           7       1
           mark iii         16      36

Single label. Note this returns a DataFrame with a single index.

>>> df.loc['cobra']
         max_speed  shield
mark i          12       2
mark ii          0       4

Single index tuple. Note this returns a Series.

>>> df.loc[('cobra', 'mark ii')]
max_speed    0
shield       4
Name: (cobra, mark ii), dtype: int64

Single label for row and column. Similar to passing in a tuple, this
returns a Series.

>>> df.loc['cobra', 'mark i']
max_speed    12
shield        2
Name: (cobra, mark i), dtype: int64

Single tuple. Note using ``[[]]`` returns a DataFrame.

>>> df.loc[[('cobra', 'mark ii')]]
               max_speed  shield
cobra mark ii          0       4

Single tuple for the index with a single label for the column

>>> df.loc[('cobra', 'mark i'), 'shield']
2

Slice from index tuple to single label

>>> df.loc[('cobra', 'mark i'):'viper']
                     max_speed  shield
cobra      mark i           12       2
           mark ii           0       4
sidewinder mark i           10      20
           mark ii           1       4
viper      mark ii           7       1
           mark iii         16      36

Slice from index tuple to index tuple

>>> df.loc[('cobra', 'mark i'):('viper', 'mark ii')]
                    max_speed  shield
cobra      mark i          12       2
           mark ii          0       4
sidewinder mark i          10      20
           mark ii          1       4
viper      mark ii          7       1

In [21]:
csEmi3 = customerData.loc[(customerData.CUSTID == 'Z12772') & (customerData.EMI == 2)]
csEmi3

CUSTID  AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
144  Z12772    3029.6    2            4    2.093035  Big-Screen-lover

In [63]:
csEmi3 = customerData.loc[customerData.EMI == 2]
csEmi3

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
1    A14219  3642.441950    2            4    2.682023      Sleeping-dog
2    A15312  3653.884565    2            5    3.208202      Sleeping-dog
3    A16605  3713.211107    2            6    0.900000        Early-bird
4    B10634  3391.074215    2            4    2.453656      Sleeping-dog
5    B16849  3755.263391    2            4    2.786665      Sleeping-dog
..      ...          ...  ...          ...         ...               ...
144  Z12772  3029.600000    2            4    2.093035  Big-Screen-lover
145  Z13253  3678.800000    2            5    2.463670  Big-Screen-lover
147  Z16428  3516.500000    2            5    2.371301  Big-Screen-lover
148  Z16735  3300.100000    2            5    1.704942  Big-Screen-lover
149  Z19835  3326.526183    2            6    1.200000        Early-bird

[120 rows x 6 columns]

### 인덱스로 조회하기

In [108]:
specificIndex = [0,3]    # 우리가 사용하는 인덱스 개념과는 다르다. loc을 사용하면 그냥 값을 사용한다. 0, 3을 가져와라.
customerData.loc[specificIndex]

CUSTID     AVGPRICE EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0  A13566  4273.900000   3            6    1.679181  Big-Screen-lover
3  A16605  3713.211107   2            6    0.900000        Early-bird

In [109]:
customerData.loc[0:3]    # 우리가 사용하는 인덱스 개념과는 다르다. loc을 사용하면 그냥 값을 사용한다. 0 ~ 3 을 다 가져와라.

CUSTID     AVGPRICE EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0  A13566  4273.900000   3            6    1.679181  Big-Screen-lover
1  A14219  3642.441950   2            4    2.682023      Sleeping-dog
2  A15312  3653.884565   2            5    3.208202      Sleeping-dog
3  A16605  3713.211107   2            6    0.900000        Early-bird

In [75]:
customerData[0:3]    # 우리가 사용하는 인덱스 개념.

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0  A13566  4273.900000    3            6    1.679181  Big-Screen-lover
1  A14219  3642.441950    2            4    2.682023      Sleeping-dog
2  A15312  3653.884565    2            5    3.208202      Sleeping-dog

In [107]:
customerData.loc[:]    # 이건 전부 다 조회를 시킨다.

CUSTID     AVGPRICE EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0    A13566  4273.900000   3            6    1.679181  Big-Screen-lover
1    A14219  3642.441950   2            4    2.682023      Sleeping-dog
2    A15312  3653.884565   2            5    3.208202      Sleeping-dog
3    A16605  3713.211107   2            6    0.900000        Early-bird
4    B10634  3391.074215   2            4    2.453656      Sleeping-dog
..      ...          ...  ..          ...         ...               ...
145  Z13253  3678.800000   2            5    2.463670  Big-Screen-lover
146  Z13534  3662.437527   3            6    0.900000        Early-bird
147  Z16428  3516.500000   2            5    2.371301  Big-Screen-lover
148  Z16735  3300.100000   2            5    1.704942  Big-Screen-lover
149  Z19835  3326.526183   2            6    1.200000        Early-bird

[150 rows x 6 columns]

In [66]:
customerData[(customerData.EMI == 2) &
             (customerData.DEVICECOUNT > 4)]

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
2    A15312  3653.884565    2            5    3.208202      Sleeping-dog
3    A16605  3713.211107    2            6    0.900000        Early-bird
6    B18816  3213.911446    2            6    0.900000        Early-bird
7    C10639  3516.500000    2            5    2.564830  Big-Screen-lover
8    C12717  3732.900000    2            5    2.069995  Big-Screen-lover
..      ...          ...  ...          ...         ...               ...
143  Z12343  3498.896985    2            6    0.900000        Early-bird
145  Z13253  3678.800000    2            5    2.463670  Big-Screen-lover
147  Z16428  3516.500000    2            5    2.371301  Big-Screen-lover
148  Z16735  3300.100000    2            5    1.704942  Big-Screen-lover
149  Z19835  3326.526183    2            6    1.200000        Early-bird

[75 rows x 6 columns]

In [97]:
customerData.loc[(customerData.EMI == 3) &
                 (customerData.CUSTTYPE == 'Early-bird')].head()

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE    CUSTTYPE
27  F17131  3179.708359    3            6         1.1  Early-bird
33  G16437  3787.328898    3            6         1.0  Early-bird
36  G18402  3402.199381    3            6         1.0  Early-bird
40  H16326  3988.935304    3            6         0.9  Early-bird
45  I12899  3018.895637    3            6         0.9  Early-bird

### 데이터프레임 특정 컬럼 데이터 타입 바꾸기

In [98]:
# 컬럼 데이터 타입 바꾸기. (저장 O)
customerData['EMI'] = customerData.EMI.astype(str)    # str, int, float
customerData.dtypes

CUSTID          object
AVGPRICE       float64
EMI             object
DEVICECOUNT      int64
PRODUCTAGE     float64
CUSTTYPE        object
dtype: object

In [87]:
customerData.loc[(customerData.EMI == '3')].head()     # 이제 '3'로 조회를 해야한다.

CUSTID     AVGPRICE EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0   A13566  4273.900000   3            6    1.679181  Big-Screen-lover
27  F17131  3179.708359   3            6    1.100000        Early-bird
33  G16437  3787.328898   3            6    1.000000        Early-bird
36  G18402  3402.199381   3            6    1.000000        Early-bird
40  H16326  3988.935304   3            6    0.900000        Early-bird

In [91]:
# 컬럼 데이터 타입 바꾸기. (저장 X) (조회를 위해 잠시 필요할 때 사용한다.)
customerData.loc[customerData.EMI.astype(int) > 2].head()

CUSTID     AVGPRICE EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0   A13566  4273.900000   3            6    1.679181  Big-Screen-lover
27  F17131  3179.708359   3            6    1.100000        Early-bird
33  G16437  3787.328898   3            6    1.000000        Early-bird
36  G18402  3402.199381   3            6    1.000000        Early-bird
40  H16326  3988.935304   3            6    0.900000        Early-bird

In [90]:
customerData.dtypes

CUSTID          object
AVGPRICE       float64
EMI             object
DEVICECOUNT      int64
PRODUCTAGE     float64
CUSTTYPE        object
dtype: object

## 숫자 & 문자 조건도 검색 되는데 굳이 왜 임시로 바꿀까?
### "201702"라는 데이터가 있다. 일반적으로 문자열 처리를 많이 하기 때문에 기본 컬럼 데이터 타입은 문자열로 두고, 이것을 계산에 사용할 때 yearweek > 201702 이런 계산이 필요할 경우 잠시 int 타입으로 바꿔서 계산한다. 데이터 자체를 바꾸기에는 문자열 데이터를 더 많이 쓰기 때문에 이렇게 사용함.

## 리스트를 통해 데이터프레임 조회하기

In [101]:
emiCondition = ['2','3']
customerData.loc[customerData.EMI.isin(emiCondition)]    # is in : emiCondition이라는 리스트에 의해 EMI가 '2', '3'인 데이터를 조회.

CUSTID     AVGPRICE EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0    A13566  4273.900000   3            6    1.679181  Big-Screen-lover
1    A14219  3642.441950   2            4    2.682023      Sleeping-dog
2    A15312  3653.884565   2            5    3.208202      Sleeping-dog
3    A16605  3713.211107   2            6    0.900000        Early-bird
4    B10634  3391.074215   2            4    2.453656      Sleeping-dog
..      ...          ...  ..          ...         ...               ...
145  Z13253  3678.800000   2            5    2.463670  Big-Screen-lover
146  Z13534  3662.437527   3            6    0.900000        Early-bird
147  Z16428  3516.500000   2            5    2.371301  Big-Screen-lover
148  Z16735  3300.100000   2            5    1.704942  Big-Screen-lover
149  Z19835  3326.526183   2            6    1.200000        Early-bird

[139 rows x 6 columns]

In [102]:
customerData.loc[~customerData.EMI.isin(emiCondition)]    # ~ 에 의해 반전된다. EMI가 '2', '3'이 아닌 데이터를 조회.

CUSTID     AVGPRICE EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
11   C18622  3246.000000   1            5    1.891624  Big-Screen-lover
46   I14835  3374.091339   1            4    2.749678      Sleeping-dog
47   I17056  2963.370122   1            3    2.139196      Sleeping-dog
56   K10712  3419.462596   1            4    2.825790      Sleeping-dog
57   K13300  3159.839204   1            3    1.960833      Sleeping-dog
65   M16452  3353.537290   1            4    1.940357      Sleeping-dog
68   N14862  2390.083736   1            6    1.000000        Early-bird
72   O13037  3398.216334   1            4    2.994395      Sleeping-dog
73   O13586  2980.229239   1            3    2.038603      Sleeping-dog
86   Q16449  3048.308860   1            3    2.369758      Sleeping-dog
140  Y12458  2976.211190   1            3    2.199677      Sleeping-dog

# 3. 데이터 다루기 (고급)

kopo_channel_seasonality_new.csv 파일을 불러온 후 selloutData 변수에 담고 QTY 컬럼 -> 실수(float), 이외 컬럼 -> 문자(str)로 변경하세요.

In [2]:
import pandas as pd

url = '../exampleCode/dataset/kopo_channel_seasonality_new.csv'
selloutData = pd.read_csv(url)
selloutData

REGIONID    PRODUCT  YEARWEEK      QTY
0           A60   PRODUCT4    201402     71.0
1           A60  PRODUCT59    201402  22275.0
2           A60  PRODUCT34    201402   4463.0
3           A60  PRODUCT47    201402      0.0
4           A60  PRODUCT56    201402     23.0
...         ...        ...       ...      ...
124653      A10  PRODUCT60    201630    824.0
124654      A10  PRODUCT56    201630    275.0
124655      A10  PRODUCT61    201630      0.0
124656      A10  PRODUCT12    201630  15021.0
124657      A10   PRODUCT1    201630    568.0

[124658 rows x 4 columns]

In [3]:
selloutData.dtypes

REGIONID     object
PRODUCT      object
YEARWEEK      int64
QTY         float64
dtype: object

In [4]:
# 기존에 float이었어도 다시 float으로 바꾼다.
# 왜냐하면, 데이터를 처리하다 타입이 바뀌었을 수도 있고, 다른 지역의 selloutData는 float이 아닐 수도 있다.
# 이렇게 처리를 하고 시작하면 내 코드는 외부 영향을 받지 않는다.

selloutData['QTY'] = selloutData.QTY.astype(float)
selloutData['REGIONID'] = selloutData.REGIONID.astype(str)
selloutData['PRODUCT'] = selloutData.PRODUCT.astype(str)
selloutData['YEARWEEK'] = selloutData.YEARWEEK.astype(str)

In [5]:
selloutData.dtypes

REGIONID     object
PRODUCT      object
YEARWEEK     object
QTY         float64
dtype: object

이거 아래쪽 인덱스로 조회하기 (고급) 위에다 옮기자...

In [7]:
url = '../exampleCode/dataset/customerdata.csv'
customerData = pd.read_csv(url)
customerData

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0    A13566  4273.900000    3            6    1.679181  Big-Screen-lover
1    A14219  3642.441950    2            4    2.682023      Sleeping-dog
2    A15312  3653.884565    2            5    3.208202      Sleeping-dog
3    A16605  3713.211107    2            6    0.900000        Early-bird
4    B10634  3391.074215    2            4    2.453656      Sleeping-dog
..      ...          ...  ...          ...         ...               ...
145  Z13253  3678.800000    2            5    2.463670  Big-Screen-lover
146  Z13534  3662.437527    3            6    0.900000        Early-bird
147  Z16428  3516.500000    2            5    2.371301  Big-Screen-lover
148  Z16735  3300.100000    2            5    1.704942  Big-Screen-lover
149  Z19835  3326.526183    2            6    1.200000        Early-bird

[150 rows x 6 columns]

In [8]:
targetColumns = ['CUSTID','AVGPRICE']
customerDataC = customerData.loc[3:5, targetColumns]    # 컬럼 명으로 조회하기.
customerDataC

CUSTID     AVGPRICE
3  A16605  3713.211107
4  B10634  3391.074215
5  B16849  3755.263391

In [9]:
customerDataC = customerData.iloc[3:5, [0,1]]   # 컬럼 인덱스로 조회하기. 0번 컬럼이 CUSTID, 1번 컬럼이 AVGPRICE다.
customerDataC                                   # iloc일 때는 인덱스로 작동하네... 3부터 5 전까지...

CUSTID     AVGPRICE
3  A16605  3713.211107
4  B10634  3391.074215

In [10]:
# iloc은 이렇게 활용할 수 있다. (나중에 컬럼 넘버가 바뀌어도 반응할 수 있게 된다!!!)
custidColumnNum = 0
avgpriceColumnNum = 1
customerDataC = customerData.iloc[3:5, [custidColumnNum,avgpriceColumnNum]]
customerDataC

CUSTID     AVGPRICE
3  A16605  3713.211107
4  B10634  3391.074215

In [11]:
# 위에 방법처럼 컬럼 넘버가 바뀌어도 무관하며 코드 가독성을 높이기 위해 요즘은 이런 추세로 바뀌는 중이다.
customerDataC = customerData.loc[3:5, ['CUSTID','AVGPRICE']]
customerDataC

CUSTID     AVGPRICE
3  A16605  3713.211107
4  B10634  3391.074215
5  B16849  3755.263391

In [12]:
customerData.loc[:, ['EMI']]

EMI
0      3
1      2
2      2
3      2
4      2
..   ...
145    2
146    3
147    2
148    2
149    2

[150 rows x 1 columns]

In [13]:
customerData.loc[:, ['EMI']].drop_duplicates()    # 중복 제거

EMI
0     3
1     2
11    1

### 인덱스 리셋

In [14]:
# 특정 행 조회로 인덱스가 깨짐.
customerDataIndex = customerData.loc[[2,4,6]]
customerDataIndex

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE      CUSTTYPE
2  A15312  3653.884565    2            5    3.208202  Sleeping-dog
4  B10634  3391.074215    2            4    2.453656  Sleeping-dog
6  B18816  3213.911446    2            6    0.900000    Early-bird

In [15]:
# 인덱스 리셋 (기존 인덱스 유지)
customerDataIndexY = customerDataIndex.reset_index()
customerDataIndexY

index  CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE      CUSTTYPE
0      2  A15312  3653.884565    2            5    3.208202  Sleeping-dog
1      4  B10634  3391.074215    2            4    2.453656  Sleeping-dog
2      6  B18816  3213.911446    2            6    0.900000    Early-bird

In [16]:
# 인덱스 리셋 (기존 인덱스 제거)
customerDataIndexN = customerDataIndex.reset_index(drop = True)
customerDataIndexN

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE      CUSTTYPE
0  A15312  3653.884565    2            5    3.208202  Sleeping-dog
1  B10634  3391.074215    2            4    2.453656  Sleeping-dog
2  B18816  3213.911446    2            6    0.900000    Early-bird

In [24]:
csEmi3 = customerData.loc[customerData.EMI == 3]
csEmi3.reset_index(drop = True).head()    # 변수에 다시 담아주지 않으면 저장이 되지 않는다.

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0  A13566  4273.900000    3            6    1.679181  Big-Screen-lover
1  F17131  3179.708359    3            6    1.100000        Early-bird
2  G16437  3787.328898    3            6    1.000000        Early-bird
3  G18402  3402.199381    3            6    1.000000        Early-bird
4  H16326  3988.935304    3            6    0.900000        Early-bird

In [26]:
csEmi3.head()

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0   A13566  4273.900000    3            6    1.679181  Big-Screen-lover
27  F17131  3179.708359    3            6    1.100000        Early-bird
33  G16437  3787.328898    3            6    1.000000        Early-bird
36  G18402  3402.199381    3            6    1.000000        Early-bird
40  H16326  3988.935304    3            6    0.900000        Early-bird

### inplace

In [30]:
csEmi3.reset_index(drop = True, inplace = True)    # 기본 inplace가 False다. True로 해주자.

In [31]:
csEmi3.head()

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
0  A13566  4273.900000    3            6    1.679181  Big-Screen-lover
1  F17131  3179.708359    3            6    1.100000        Early-bird
2  G16437  3787.328898    3            6    1.000000        Early-bird
3  G18402  3402.199381    3            6    1.000000        Early-bird
4  H16326  3988.935304    3            6    0.900000        Early-bird

In [17]:
# 특정 컬럼을 인덱스로 만들기
customerDataNewIndex = customerData.set_index(keys = ['CUSTID'])
customerDataNewIndex

AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
CUSTID                                                             
A13566  4273.900000    3            6    1.679181  Big-Screen-lover
A14219  3642.441950    2            4    2.682023      Sleeping-dog
A15312  3653.884565    2            5    3.208202      Sleeping-dog
A16605  3713.211107    2            6    0.900000        Early-bird
B10634  3391.074215    2            4    2.453656      Sleeping-dog
...             ...  ...          ...         ...               ...
Z13253  3678.800000    2            5    2.463670  Big-Screen-lover
Z13534  3662.437527    3            6    0.900000        Early-bird
Z16428  3516.500000    2            5    2.371301  Big-Screen-lover
Z16735  3300.100000    2            5    1.704942  Big-Screen-lover
Z19835  3326.526183    2            6    1.200000        Early-bird

[150 rows x 5 columns]

In [18]:
customerDataNewIndex = customerData.set_index(keys = ['CUSTTYPE'])
customerDataNewIndex

CUSTID     AVGPRICE  EMI  DEVICECOUNT  PRODUCTAGE
CUSTTYPE                                                           
Big-Screen-lover  A13566  4273.900000    3            6    1.679181
Sleeping-dog      A14219  3642.441950    2            4    2.682023
Sleeping-dog      A15312  3653.884565    2            5    3.208202
Early-bird        A16605  3713.211107    2            6    0.900000
Sleeping-dog      B10634  3391.074215    2            4    2.453656
...                  ...          ...  ...          ...         ...
Big-Screen-lover  Z13253  3678.800000    2            5    2.463670
Early-bird        Z13534  3662.437527    3            6    0.900000
Big-Screen-lover  Z16428  3516.500000    2            5    2.371301
Big-Screen-lover  Z16735  3300.100000    2            5    1.704942
Early-bird        Z19835  3326.526183    2            6    1.200000

[150 rows x 5 columns]

In [19]:
customerDataNewIndex = customerData.set_index(keys = ['EMI'])
customerDataNewIndex

CUSTID     AVGPRICE  DEVICECOUNT  PRODUCTAGE          CUSTTYPE
EMI                                                                
3    A13566  4273.900000            6    1.679181  Big-Screen-lover
2    A14219  3642.441950            4    2.682023      Sleeping-dog
2    A15312  3653.884565            5    3.208202      Sleeping-dog
2    A16605  3713.211107            6    0.900000        Early-bird
2    B10634  3391.074215            4    2.453656      Sleeping-dog
..      ...          ...          ...         ...               ...
2    Z13253  3678.800000            5    2.463670  Big-Screen-lover
3    Z13534  3662.437527            6    0.900000        Early-bird
2    Z16428  3516.500000            5    2.371301  Big-Screen-lover
2    Z16735  3300.100000            5    1.704942  Big-Screen-lover
2    Z19835  3326.526183            6    1.200000        Early-bird

[150 rows x 5 columns]

In [20]:
customerDataIndex.info()    # 데이터 형태에는 변화가 없다

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3 entries, 2 to 6
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   CUSTID       3 non-null      object 
 1   AVGPRICE     3 non-null      float64
 2   EMI          3 non-null      int64  
 3   DEVICECOUNT  3 non-null      int64  
 4   PRODUCTAGE   3 non-null      float64
 5   CUSTTYPE     3 non-null      object 
dtypes: float64(2), int64(2), object(2)
memory usage: 168.0+ bytes


### 5-1 데이터 분석 및 데이터 전처리 P.49

### kopo_customerdata.csv 파일을 불러온 후 kopo_customerData 변수에 담고, STATENAME, GENDER 컬럼 첫 5개 행을 조회하세요.

In [33]:
import pandas as pd

url = '../exampleCode/dataset/kopo_customerdata.csv'
kopo_customerData = pd.read_csv(url)
kopo_customerData

CUSTOMERCODE STATENAME  ST  GENDER  DOB  GENDER1  EMAIL  FEST_CNT  \
0            1503989    State2   2    Male    0        1      1         0   
1            1190338    State2   2    Male    0        1      0         1   
2            1424715    State2   2  Female    0        2      0         0   
3            2483305    State2   2    Male    0        1      0         1   
4            1178802    State2   2    Male    0        1      0         1   
...              ...       ...  ..     ...  ...      ...    ...       ...   
299995        465757    State1   1    Male    0        1      0         0   
299996       1890264    State1   1    Male    0        1      0         1   
299997       1857537    State1   1  Female    0        2      0         0   
299998       2209324    State1   1    Male    0        1      0         1   
299999       1875146    State1   1    Male    0        1      0         1   

        TOTAL_AMOUNT  AC_AMOUNT  AV_AMOUNT  HA_AMOUNT  EMI_CE_AMOUNT  \
0              30300          0      30300          0              0   
1              46500          0      46500          0              0   
2              22700          0          0      22700              0   
3              31400          0      31400          0              0   
4              44900          0          0      44900              0   
...              ...        ...        ...        ...            ...   
299995         40600          0      40600          0              0   
299996         48000      48000          0          0              0   
299997         32300          0          0      32300              0   
299998         48500          0      48500          0              0   
299999         16500          0          0      16500              0   

        IS_HIGHEND  CNT_VIST  GAP_VISIT  
0                0         1          0  
1                0         1          0  
2                0         1          0  
3                0         1          0  
4                0         1          0  
...            ...       ...        ...  
299995           0         1          0  
299996           0         1          0  
299997           0         1          0  
299998           0         1          0  
299999           0         1          0  

[300000 rows x 16 columns]

In [63]:
kopo_customerData.loc[0:4,['STATENAME','GENDER']].reset_index(drop = True)

STATENAME  GENDER
0    State2    Male
1    State2    Male
2    State2  Female
3    State2    Male
4    State2    Male

In [62]:
targetColumns = ['STATENAME','GENDER']
kopo_customerData.loc[0:4, targetColumns].reset_index(drop = True)

STATENAME  GENDER
0    State2    Male
1    State2    Male
2    State2  Female
3    State2    Male
4    State2    Male

In [60]:
kopo_customerData.loc[:,['STATENAME','GENDER']].reset_index(drop = True).head()

STATENAME  GENDER
0    State2    Male
1    State2    Male
2    State2  Female
3    State2    Male
4    State2    Male

In [61]:
kopo_customerData[['STATENAME','GENDER']].reset_index(drop = True).head()

STATENAME  GENDER
0    State2    Male
1    State2    Male
2    State2  Female
3    State2    Male
4    State2    Male

In [64]:
statenameColNum = 1
gendercolNum = 3
kopo_customerData.iloc[0:5, [statenameColNum,gendercolNum]].reset_index(drop = True)

STATENAME  GENDER
0    State2    Male
1    State2    Male
2    State2  Female
3    State2    Male
4    State2    Male